### PR TITLE
AA-45 Refactor handle ops

### PR DESCRIPTION
**Pull request by drortirosh**

### CODE REVIEW

File: contracts/EntryPoint.sol [hunk 1]
```diff
@@ -83,6 +83,27 @@ contract EntryPoint is StakeManager {
...
+     * @param opIndex into into the opInfo array
+     * @param userOp the userOp to execute
+     * @param opInfo the opInfo filled by validatePrepayment for this userOp.
+     * @return collected the total amount this userOp paid.
+     */
+    function _executeUserOp(uint256 opIndex, UserOperation calldata userOp, UserOpInfo memory opInfo) private returns (uint256 collected) {
+        uint256 preGas = gasleft();
+        uint256 contextOffset = opInfo.contextOffset;
+        bytes memory context;
+        assembly {context := contextOffset}
```
Comment (forshtat) [inline contracts/EntryPoint.sol]: Can we convert these assembly pieces into utility functions of
`function getUintMemoryPointer(bytes memory) private view returns (uint)`
and
`function getBytesFromMemoryPointer(uint point) private view returns (bytes memory)`
?
Like, Solidity won't copy the `bytes memory` to pass it into a function, right?

Not that it is necessary, just weird one-liner assembly insertions in the middle of regular function seems weird.

File: contracts/EntryPoint.sol [hunk 2]
```diff
@@ -83,6 +83,27 @@ contract EntryPoint is StakeManager {
...
+
+        try this.innerHandleOp(userOp, opInfo, context) returns (uint256 _actualGasCost) {
+            collected = _actualGasCost;
+        } catch {
+            uint256 actualGas = preGas - gasleft() + opInfo.preOpGas;
```
Comment (forshtat) [inline contracts/EntryPoint.sol:95]: I understand that `preOpGas` is 'gas used before execution'?

Let me just point out that we have in this code:

- `verificationGas` which is a limit on gas for verification
- `preVerificationGas` which is an exact value for gas overhead to be added
- `preOpGas` which is a partial gas used value

I *strongly suggest* we spend some more letters in the names and call these `XxxLimit`, `XxxUsed`, `XxxOverhead` etc.

Comment (drortirosh) [reply contracts/EntryPoint.sol:95]: ok. but its a shotgun renaming. I'll add it in a separate PR.

File: contracts/EntryPoint.sol [hunk 3]
```diff
@@ -139,7 +132,7 @@ contract EntryPoint is StakeManager {
     struct UserOpInfo {
         bytes32 requestId;
         uint256 prefund;
-        PaymentMode paymentMode;
+        address paymaster;
```
Comment (forshtat) [inline contracts/EntryPoint.sol]: `PaymentMode` is not used any more

File: contracts/EntryPoint.sol [hunk 4]
```diff
@@ -191,30 +185,35 @@ contract EntryPoint is StakeManager {
...
-    function _getPaymentInfo(UserOperation calldata userOp) internal view returns (uint256 requiredPrefund, PaymentMode paymentMode) {
-        requiredPrefund = userOp.requiredPreFund();
-        if (userOp.hasPaymaster()) {
-            paymentMode = PaymentMode.paymasterDeposit;
-        } else {
-            paymentMode = PaymentMode.walletDeposit;
-        }
+    function _getPaymentInfo(UserOperation calldata userOp) internal view returns (uint256 requiredPrefund, address paymaster) {
+    unchecked {
+        paymaster = userOp.paymaster;
```
Comment (forshtat) [inline contracts/EntryPoint.sol]: There seems to be no point in returning `address paymaster` unmodified here.

File: contracts/EntryPoint.sol [hunk 5]
```diff
@@ -191,30 +185,35 @@ contract EntryPoint is StakeManager {
...
+        //when using a Paymaster, the verificationGas is used also to as a limit for the postOp call.
+        // our security model might call postOp eventually twice
+        uint256 mul = paymaster != address(0) ? 3 : 1;
```
Comment (forshtat) [inline contracts/EntryPoint.sol]: I am having my doubts that implicit `postOpGasLimit := verificationGasLimit` is such a great assumption.
IIRC, wallet deployment cost is counted towards the verification gas, isn't it? 
This potentially means much bigger then expected `requiredGas`, and way too much gas available to the `postOp`

File: contracts/EntryPoint.sol [hunk 6]
```diff
@@ -281,9 +280,8 @@ contract EntryPoint is StakeManager {
      * revert with proper FailedOp in case paymaster reverts.
      * decrement paymaster's deposit
      */
-    function _validatePaymasterPrepayment(uint256 opIndex, UserOperation calldata op, bytes32 requestId, uint256 requiredPreFund, uint256 gasUsedByValidateWalletPrepayment) internal returns (bytes memory context) {
+    function _validatePaymasterPrepayment(uint256 opIndex, address paymaster, UserOperation calldata op, bytes32 requestId, uint256 requiredPreFund, uint256 gasUsedByValidateWalletPrepayment) internal returns (bytes memory context) {
```
Comment (forshtat) [inline contracts/EntryPoint.sol]: What is the point of doing it this way?

File: contracts/EntryPoint.sol (hunk 3 above)
Comment (forshtat) [inline contracts/EntryPoint.sol]: Also, I don't think there is a single function here that has `UserOpInfo` but does not have a corresponding `UserOperation`, making this field a duplicate.

File: contracts/EntryPoint.sol [hunk 7]
```diff
@@ -310,24 +308,27 @@ contract EntryPoint is StakeManager {
      * also make sure total validation doesn't exceed verificationGas
      * this method is called off-chain (simulateValidation()) and on-chain (from handleOps)
      */
-    function _validatePrepayment(uint256 opIndex, UserOperation calldata userOp, bytes32 requestId) private returns (uint256 requiredPreFund, PaymentMode paymentMode, bytes memory context){
+    function _validatePrepayment(uint256 opIndex, UserOperation calldata userOp, UserOpInfo[] memory opInfos) private {
```
Comment (forshtat) [inline contracts/EntryPoint.sol]: Just a question - can we pass an individual `UserOpInfo` here? I feel worried that this function accepts the entire array just to set a few value in a single element. Even if gas-wise this is cheap.

File: contracts/EntryPoint.sol (hunk 3 above)
Comment (drortirosh) [reply contracts/EntryPoint.sol]: yes, it is not needed: paymaster!=address(0) means "paymaster pays".
(it used to be a 3-state, but now the mode is boolean)

Comment (drortirosh) [reply contracts/EntryPoint.sol]: it isn't. its an in-memory collected data about a userOp (context, prefund, etc)
it also caches "paymaster" address.

File: contracts/EntryPoint.sol (hunk 4 above)
Comment (drortirosh) [reply contracts/EntryPoint.sol]: I use the return value to cache into the opInfo

File: contracts/EntryPoint.sol (hunk 7 above)
Comment (drortirosh) [reply contracts/EntryPoint.sol]: no, you can't.
even making the assignment as:
```
        UserOpInfo memory opInfo = opInfos[opIndex];
        opInfo = UserOpInfo(...)
```
causes a revert on geth (and interestingly, hardhat is fully stuck..)
a "reference" into an array is not assignable

File: contracts/EntryPoint.sol (hunk 6 above)
Comment (drortirosh) [reply contracts/EntryPoint.sol]: avoiding reading the paymaster again from the calldata. I read it once (to know the payment mode) and then use it from memory. cheaper to pass it as parameter than reading it again from calldata. (solidity manages to cost ~200gas for each calldata member. it does a lot of useless boundary checks)

File: contracts/EntryPoint.sol (hunk 3 above)
Comment (forshtat) [reply contracts/EntryPoint.sol]: I mean, you should also remove the `enum` declaration.

File: contracts/EntryPoint.sol (hunk 7 above)
Comment (forshtat) [reply contracts/EntryPoint.sol]: Do you need to assign to 'reference'? It is already pointing somewhere, can't you just assign every field of a struct?

```
UserOpInfo memory opInfo = opInfos[opIndex];
opInfo.paymaster = paymaster;
opInfo.prepaid = prepaid;
...
```
?

File: contracts/EntryPoint.sol (hunk 5 above)
Comment (drortirosh) [reply contracts/EntryPoint.sol]: we do need SOME limit on postOp - we need verify the paymaster has "max possible gas" to pay for it.
The options are:
- constant value
- add new `UserOp.postOpGasLimit` parameter (which is used only if there's a paymaster)
- verificationGas
- callGas
- a factor (e.g. half) of one of the above.

using verificationGas is not ideal, but better than the alternatives. The only downside is that the paymaster needs a higher balance (which is never used)

**Pull request merged**